src/Core/Cart/CartRuleCalculator.php

```diff
@@ -122,65 +122,67 @@ protected function applyCartRule(CartRuleData $cartRuleData, $withFreeShipping =
+                    if (((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
+                            || !$cartRule->reduction_exclude_special)) && ($cartRowCheapest === null
+                            || $cartRowCheapest->getInitialUnitPrice()->getTaxIncluded() > $cartRow->getInitialUnitPrice()
+                                ->getTaxIncluded())
+                    ) {
+                        $cartRowCheapest = $cartRow;
+                    }
+                }
+                if ($cartRowCheapest !== null) {
+                    // apply only on one product of the cheapest row
+                    $discountTaxIncluded = $cartRowCheapest->getInitialUnitPrice()->getTaxIncluded()
+                        * $cartRule->reduction_percent / 100;
+                    $discountTaxExcluded = $cartRowCheapest->getInitialUnitPrice()->getTaxExcluded()
+                        * $cartRule->reduction_percent / 100;
+                    $amount = new AmountImmutable($discountTaxIncluded, $discountTaxExcluded);
+                    $cartRowCheapest->applyFlatDiscount($amount);
+                    $cartRuleData->addDiscountApplied($amount);
                 }
             }
-            if ($cartRowCheapest !== null) {
-                // apply only on one product of the cheapest row
-                $discountTaxIncluded = $cartRowCheapest->getInitialUnitPrice()->getTaxIncluded()
-                                       * $cartRule->reduction_percent / 100;
-                $discountTaxExcluded = $cartRowCheapest->getInitialUnitPrice()->getTaxExcluded()
-                                       * $cartRule->reduction_percent / 100;
-                $amount = new AmountImmutable($discountTaxIncluded, $discountTaxExcluded);
-                $cartRowCheapest->applyFlatDiscount($amount);
-                $cartRuleData->addDiscountApplied($amount);
-            }
-        }
 
-        // Discount (%) on the selection of products
-        if ($cartRule->reduction_percent && $cartRule->reduction_product == -2) {
-            $selected_products = $cartRule->checkProductRestrictionsFromCart($cart, true);
-            if (is_array($selected_products)) {
-                foreach ($this->cartRows as $cartRow) {
-                    $product = $cartRow->getRowData();
-                    if (in_array($product['id_product'] . '-' . $product['id_product_attribute'], $selected_products)
-                        || in_array($product['id_product'] . '-0', $selected_products)
-                           && (($cartRule->reduction_exclude_special && !$product['reduction_applies'])
-                               || !$cartRule->reduction_exclude_special)) {
-                        $amount = $cartRow->applyPercentageDiscount($cartRule->reduction_percent);
-                        $cartRuleData->addDiscountApplied($amount);
+            // Discount (%) on the selection of products
+            if ($cartRule->reduction_product == -2) {
```

Comment:
Can we improve a little bit the code and replace this `-2` by something that has meaning, like either a variable or a constant ?